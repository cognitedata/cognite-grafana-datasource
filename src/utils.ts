import { isNil, omitBy, get } from 'lodash';
import { stringify } from 'query-string';
import ms from 'ms';
import { QueryOptions, QueryTarget } from './types';
import { FilterTypes, ParsedFilter } from './parser/types';

export function getQueryString(obj: any) {
  return stringify(omitBy(obj, isNil));
}

export function toGranularityWithLowerBound(milliseconds: number, lowerBound = 1000): string {
  return ms(Math.max(milliseconds, lowerBound));
}

// used for generating the options.requestId
export function getRequestId(options: QueryOptions, target: QueryTarget) {
  return `${options.dashboardId}_${options.panelId}_${target.refId}`;
}

export const applyFilters = <T>(objs: T[], filters: ParsedFilter[]): T[] => {
  if (!filters.length) {
    return objs;
  }

  return objs.filter((obj) => filters.every((filter) => checkFilter(obj, filter)));
};

export const checkFilter = <T>(obj: T, { path, filter, value }: ParsedFilter): boolean => {
  const valueToFilter = get(obj, path, null);
  const regex = new RegExp(`^${value}$`);

  switch (filter) {
    case FilterTypes.RegexEquals:
      return regex.test(valueToFilter);
    case FilterTypes.RegexNotEquals:
      return !regex.test(valueToFilter);
    case FilterTypes.NotEquals:
      return value !== valueToFilter;
    default:
      return false;
  }
};
