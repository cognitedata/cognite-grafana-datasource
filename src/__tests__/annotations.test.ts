import * as _ from 'lodash';
import { getMockedDataSource } from './utils';

jest.mock('../cache');

const { ds, backendSrvMock } = getMockedDataSource();

describe('Annotations Query', () => {
  const annotationResponse = {
    data: {
      items: [
        {
          id: 1,
          assetIds: [123, 456, 789],
          description: 'event 1',
          startTime: '1549336675000',
          endTime: '1549336775000',
          type: 'type 1',
          subtype: 'subtype 1',
        },
        {
          id: 2,
          assetIds: [123],
          description: 'event 2',
          startTime: '1549336775000',
          endTime: '1549336875000',
          type: 'type 2',
          subtype: 'subtype 2',
        },
        {
          id: 3,
          assetIds: [456],
          description: 'event 3',
          startTime: '1549336875000',
          endTime: '1549336975000',
          type: 'type 3',
          subtype: 'subtype 3',
        },
        {
          id: 4,
          assetIds: [789],
          description: 'event 4',
          startTime: '1549336975000',
          endTime: '1549337075000',
          type: 'type 4',
          subtype: 'subtype 4',
        },
        {
          id: 5,
          assetIds: [123, 456, 789],
          description: 'time out of bounds',
          startTime: '1549336600000',
          endTime: '1549338500000',
          type: 'type 1',
          subtype: 'subtype 2',
          metadata: { key1: 'value1', key2: 'value2' },
        },
      ],
    },
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Given an empty annotation query', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: '',
      },
    };

    beforeAll(async () => {
      result = await ds.annotationQuery(annotationOption);
    });

    it('should return nothing', () => {
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
      expect(result).toEqual([]);
    });
  });

  describe('Given an annotation query without any filters', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: 'events{}',
      },
    };

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(annotationResponse));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should return all events', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(result.length).toEqual(annotationResponse.data.items.length);
    });
  });

  describe('Given an annotation query where no events are returned', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{type='type 5'}",
      },
    };

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve({ data: { items: [] } }));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should return empty array', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(result).toEqual([]);
    });
  });

  describe('Given an annotation query', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{assetIds=[123], type='type 1'}",
      },
    };
    const response = _.cloneDeep(annotationResponse);
    response.data.items = annotationResponse.data.items.filter(
      item => item.assetIds.some(id => id === 123) && item.type === 'type 1'
    );

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(response));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct events', () => {
      const resultIds = result.map(({ text }) => text);
      const expectedEvents = ['event 1', 'time out of bounds'];

      expect(result.length).toEqual(2);
      expect(expectedEvents.every(text => resultIds.includes(text))).toBeTruthy();
    });
  });

  describe('Given an annotation query with a metadata request', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{metadata={key1='value1', key2='value2'}}",
      },
    };
    const response = _.cloneDeep(annotationResponse);
    response.data.items = annotationResponse.data.items.filter(item => item.metadata);

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(response));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct event', () => {
      const resultIds = result.map(({ text }) => text);

      expect(result.length).toEqual(1);
      expect(resultIds.includes('time out of bounds')).toBeTruthy();
    });
  });

  describe('Given an annotation query where nothing is returned', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{type='non-existant type'}",
      },
    };
    const response = _.cloneDeep(annotationResponse);
    response.data.items = [];

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(response));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct events', () => {
      expect(result).toEqual([]);
    });
  });

  describe('Given an annotation query with filters', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{description=~'event.*', type!='type 1'}",
      },
    };

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(annotationResponse));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct events', () => {
      const resultIds = result.map(({ text }) => text);
      const expectedEvents = ['event 2', 'event 3', 'event 4'];

      expect(result.length).toEqual(3);
      expect(expectedEvents.every(text => resultIds.includes(text))).toBeTruthy();
    });
  });

  describe('Given an annotation query with variables', () => {
    let result;
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: "events{assetIds=[$AssetVariable], description!~'event.*'}",
      },
    };
    const response = _.cloneDeep(annotationResponse);
    response.data.items = annotationResponse.data.items.filter(item =>
      item.assetIds.some(id => id === 123)
    );

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(annotationResponse));
      result = await ds.annotationQuery(annotationOption);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct events', () => {
      const resultIds = result.map(({ text }) => text);

      expect(result.length).toEqual(1);
      expect(resultIds.includes('time out of bounds')).toBeTruthy();
    });
  });

  describe('Given an annotation query with an incomplete event expression', () => {
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: 'events{ ',
      },
    };
    beforeAll(async () => {
      backendSrvMock.datasourceRequest.mockReset();
    });
    it('should throw a parse error', () => {
      expect(ds.annotationQuery(annotationOption)).rejects.toThrowErrorMatchingSnapshot();
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
    });
  });

  describe('Given an annotation query with an incorrect event expression', () => {
    const annotationOption: any = {
      range: {
        from: '1549336675000',
        to: '1549338475000',
      },
      annotation: {
        query: 'events{ name=~event, foo}',
      },
    };
    it('should throw a parse error', () => {
      expect(ds.annotationQuery(annotationOption)).rejects.toThrowErrorMatchingSnapshot();
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
    });
  });
});
