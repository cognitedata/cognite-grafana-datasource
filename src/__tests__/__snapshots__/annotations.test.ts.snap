// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Annotations Query Given an annotation query should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        123,
      ],
      "endTime": Object {
        "min": 1549336675000,
      },
      "startTime": Object {
        "max": 1549338475000,
      },
      "type": "type 1",
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;

exports[`Annotations Query Given an annotation query where nothing is returned should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "endTime": Object {
        "min": 1549336675000,
      },
      "startTime": Object {
        "max": 1549338475000,
      },
      "type": "non-existant type",
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;

exports[`Annotations Query Given an annotation query with a metadata request should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "endTime": Object {
        "min": 1549336675000,
      },
      "metadata": Object {
        "key1": "value1",
        "key2": "value2",
      },
      "startTime": Object {
        "max": 1549338475000,
      },
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;

exports[`Annotations Query Given an annotation query with an incomplete event expression should throw a parse error 1`] = `
"Parser: Unexpected end of input:
events{
       ^"
`;

exports[`Annotations Query Given an annotation query with an incorrect event expression should throw a parse error 1`] = `
"Parser: Syntax error:
events{ name=~event, foo}
              ^"
`;

exports[`Annotations Query Given an annotation query with filters should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "endTime": Object {
        "min": 1549336675000,
      },
      "startTime": Object {
        "max": 1549338475000,
      },
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;

exports[`Annotations Query Given an annotation query with variables should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        123,
      ],
      "endTime": Object {
        "min": 1549336675000,
      },
      "startTime": Object {
        "max": 1549338475000,
      },
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;

exports[`Annotations Query Given an annotation query with variables should generate the correct request 2`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        123,
        456,
      ],
      "endTime": Object {
        "min": 1549336675000,
      },
      "startTime": Object {
        "max": 1549338475000,
      },
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/events/list",
}
`;
